File: apps/backend/src/dog-meetups/dog-meetups.controller.ts
import { Body, Controller, Delete, Get, Param, Post, Query, Req, UseGuards } from '@nestjs/common';
import { DogMeetupsService } from './dog-meetups.service';
import { CreateMeetupDto } from './dto/create-meetup.dto';
import { JoinMeetupDto } from './dto/join-meetup.dto';
import { JwtAuthGuard } from '../auth/jwt-auth.guard';
import type { RequestUser } from '../auth/jwt.strategy';

@Controller('dog-meetups')
@UseGuards(JwtAuthGuard)
export class DogMeetupsController {
  constructor(private dogMeetupsService: DogMeetupsService) {}

  @Post()
  create(@Body() dto: CreateMeetupDto, @Req() req: { user: RequestUser }) {
    return this.dogMeetupsService.create(req.user.userId, dto);
  }

  @Get('nearby')
  getNearby(
    @Req() req: { user: RequestUser },
    @Query('radius') radius = '10',
    @Query('page') page = '1',
    @Query('limit') limit = '20',
  ) {
    return this.dogMeetupsService.getNearby(
      req.user.userId,
      Math.min(parseInt(radius), 50),
      parseInt(page),
      parseInt(limit),
    );
  }

  @Get(':meetupId')
  getOne(@Param('meetupId') meetupId: string) {
    return this.dogMeetupsService.getOne(meetupId);
  }

  @Delete(':meetupId')
  cancel(@Param('meetupId') meetupId: string, @Req() req: { user: RequestUser }) {
    return this.dogMeetupsService.cancel(meetupId, req.user.userId);
  }

  @Post(':meetupId/join')
  join(@Param('meetupId') meetupId: string, @Body() dto: JoinMeetupDto, @Req() req: { user: RequestUser }) {
    return this.dogMeetupsService.join(meetupId, req.user.userId, dto.petId);
  }

  @Delete(':meetupId/leave')
  leave(@Param('meetupId') meetupId: string, @Req() req: { user: RequestUser }) {
    return this.dogMeetupsService.leave(meetupId, req.user.userId);
  }
}

File: apps/backend-java/src/main/java/com/pawpals/backend/dogmeetup/DogMeetupsController.java
package com.pawpals.backend.dogmeetup;

import com.pawpals.backend.auth.RequestUser;
import com.pawpals.backend.dogmeetup.dto.CreateMeetupRequest;
import com.pawpals.backend.dogmeetup.dto.JoinMeetupRequest;
import com.pawpals.backend.dogmeetup.dto.MeetupResponse;
import jakarta.validation.Valid;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/dog-meetups")
@RequiredArgsConstructor
public class DogMeetupsController {

  private final DogMeetupsService dogMeetupsService;

  @PostMapping
  public MeetupResponse create(@Valid @RequestBody CreateMeetupRequest dto, @AuthenticationPrincipal RequestUser user) {
    return dogMeetupsService.create(user.userId(), dto);
  }

  @GetMapping("/nearby")
  public List<MeetupResponse> getNearby(
      @AuthenticationPrincipal RequestUser user,
      @RequestParam(defaultValue = "10") int radius,
      @RequestParam(defaultValue = "1") int page,
      @RequestParam(defaultValue = "20") int limit) {
    return dogMeetupsService.getNearby(user.userId(), Math.min(radius, 50), page, limit);
  }

  @GetMapping("/{meetupId}")
  public MeetupResponse getOne(@PathVariable String meetupId) {
    return dogMeetupsService.getOne(meetupId);
  }

  @DeleteMapping("/{meetupId}")
  public MeetupResponse cancel(@PathVariable String meetupId, @AuthenticationPrincipal RequestUser user) {
    return dogMeetupsService.cancel(meetupId, user.userId());
  }

  @PostMapping("/{meetupId}/join")
  public void join(
      @PathVariable String meetupId, @RequestBody(required = false) JoinMeetupRequest dto, @AuthenticationPrincipal RequestUser user) {
    dogMeetupsService.join(meetupId, user.userId(), dto != null ? dto.petId() : null);
  }

  @DeleteMapping("/{meetupId}/leave")
  public void leave(@PathVariable String meetupId, @AuthenticationPrincipal RequestUser user) {
    dogMeetupsService.leave(meetupId, user.userId());
  }
}
